macros/generic_tests/equality.sql

```diff
@@ -19,20 +23,108 @@
     {{ return('') }}
 {% endif %}
 
+
+
 -- setup
 {%- do dbt_utils._is_relation(model, 'test_equality') -%}
 
-{#-
-If the compare_cols arg is provided, we can run this test without querying the
-information schema — this allows the model to be an ephemeral model
--#}
-
+{# Ensure there are no extra columns in the compare_model vs model #}
 {%- if not compare_columns -%}
     {%- do dbt_utils._is_ephemeral(model, 'test_equality') -%}
-    {%- set compare_columns = adapter.get_columns_in_relation(model) | map(attribute='quoted') -%}
-{%- endif -%}
+    {%- do dbt_utils._is_ephemeral(compare_model, 'test_equality') -%}
+
+    {%- set model_columns = adapter.get_columns_in_relation(model) -%}
+    {%- set compare_model_columns = adapter.get_columns_in_relation(compare_model) -%}
+
+
+    {%- if exclude_columns -%}
+        {#-- Lower case ignore columns for easier comparison --#}
+        {%- set exclude_columns = exclude_columns | map("lower") | list %}
+
+        {# Filter out the excluded columns #}
+        {%- set include_columns = [] %}
+        {%- set include_model_columns = [] %}
+        {%- for column in model_columns -%}
+            {%- if column.name | lower not in exclude_columns -%}
+                {% do include_columns.append(column) %}
+            {%- endif %}
+        {%- endfor %}
+        {%- for column in compare_model_columns -%}
+            {%- if column.name | lower not in exclude_columns -%}
+                {% do include_model_columns.append(column) %}
+            {%- endif %}
+        {%- endfor %}
+
+        {%- set compare_columns_set = set(include_columns | map(attribute='quoted') | map("lower")) %}
+        {%- set compare_model_columns_set = set(include_model_columns | map(attribute='quoted') | map("lower")) %}
+    {%- else -%}
+        {%- set compare_columns_set = set(model_columns | map(attribute='quoted') | map("lower")) %}
+        {%- set compare_model_columns_set = set(compare_model_columns | map(attribute='quoted') | map("lower")) %}
+    {%- endif -%}
+
+    {% if compare_columns_set != compare_model_columns_set %}
+        {{ exceptions.raise_compiler_error(compare_model ~" has less columns than " ~ model ~ ", please ensure they have the same columns or use the `compare_columns` or `exclude_columns` arguments to subset them.") }}
+    {% endif %}
 
-{% set compare_cols_csv = compare_columns | join(', ') %}
+
+{% endif %}
+
+{%- if not precision -%}
+    {#-
+        If the compare_cols arg is provided, we can run this test without querying the
+        information schema — this allows the model to be an ephemeral model
+    -#}
```

Comment:
I think the wording and placement of this comment and the one below it are a little confusing. It makes sense if you really walk through and think about it, and I agree it's helpful to have some comments here to clarify the conditions, but you might consider a second pass at these. Not a blocking problem in my mind but wanted to flag it as they threw me at first.

Reply:
Thanks, have tried to provide some better comments!